---
title: "Faizo Kasule R assignment"
author: "Faizo"
date: "2025-03-20"
output: html_document
---

```{r}
# Clear environment
rm(list = ls())

# Closes the current graphics device
dev.off()
```


```{r}
# Set working directory (update this path to your directory)
setwd("C:/Users/DELL/OneDrive/Desktop/bcb546/EEOB lesson")

# Load required libraries
library(dplyr)
library(tidyr)
library(readr)
library(tidyverse)

# Read genotype and SNP position data
fang_et_al_genotypes <- read.table("fang_et_al_genotypes.txt", sep = "\t", header = TRUE)
snp_position <- read.table("snp_position.txt", sep = "\t", header = TRUE)
```

## Data Inspection

The following set of code inspects the data

````{r}
### This shows the dimension of the dataframe 
dim(fang_et_al_genotypes) 
dim(snp_position)

### This shows the summary of each data set 
colnames(fang_et_al_genotypes)
colnames(snp_position)
sapply(fang_et_al_genotypes, class)
sapply(snp_position, class)
is.data.frame(fang_et_al_genotypes)
is.data.frame(snp_position)
file.info("fang_et_al_genotypes")
file.info("snp_position.txt")
sum(is.na(fang_et_al_genotypes))
sum(is.na(snp_position))
cor(snp_position[sapply(snp_position, is.numeric)])
any(is.na(fang_et_al_genotypes))
any(is.na(snp_position))
any(duplicated(fang_et_al_genotypes))
any(duplicated(snp_position))
```

# Data processing 
```{r}
### This line of code copies out the data based on the group into a new file
# Maize_data and teosinte_data
maize_data_group <- fang_et_al_genotypes[fang_et_al_genotypes$Group %in% c("ZMMIL", "ZMMLR", "ZMMMR"), ]
teosinte_data_group <- fang_et_al_genotypes[fang_et_al_genotypes$Group %in% c("ZMPBA", "ZMPIL", "ZMPJA"), ]
```

```{r}
# This line of code will arrange the snp_position data frame by SNP_ID
snp_position_Chr <- snp_position[,-c(2,5:15)] 
snp_position_arranged <- arrange(snp_position_Chr, SNP_ID)
```

```{r}
library(dplyr)

# Remove columns 1 to 3 from the maize_data_group and teosinte_data_group data frame
maize_data_group <- select(maize_data_group, -(1:3))
teosinte_data_group <- select(teosinte_data_group, -(1:3))
```

# Data transposition
```{r}
# The data will be transposed using this line of code
maize_genotypes_trans <- t(maize_data_group)
teosinte_genotypes_trans <- t(teosinte_data_group)
```

# Merge SNP and transposed Genotype data
```{r}
# Combining snp_position_arranged with teosinte_genotypes_trans at the 1,1 position
teosinte_snp_genotype <- cbind(snp_position_arranged, teosinte_genotypes_trans)

# Combining snp_position_arranged with maize_genotypes_trans at the 1,1 position
maize_snp_genotype <- cbind(snp_position_arranged, maize_genotypes_trans)

# Print the number of rows for each data frame
nrow(maize_genotypes_trans)
nrow(teosinte_genotypes_trans)
nrow(snp_position_arranged)
```

# Deal with missing data
```{r}
#This replaces all the ? data with -
maize_snp_genotype <- maize_snp_genotype %>% mutate_all(function(x) gsub("\\?", "\\-",x))
teosinte_snp_genotype <- teosinte_snp_genotype %>% mutate_all(function(x) gsub("\\?", "\\-",x))
```

# Generate Maize 20 files
```{r}
#For maize
library(dplyr)
library(readr)
for (i in 1:10) {
  print(paste("Processing Chromosome", i))
  ascending_data <- maize_snp_genotype %>% 
    filter(Chromosome == i) %>% 
    arrange(as.numeric(Position))
  
  descending_data <- maize_snp_genotype %>% 
    filter(Chromosome == i) %>% 
    arrange(desc(as.numeric(Position)))
  
  
  write_tsv(ascending_data, paste("Maize_chr", i, "ascending.txt", sep = "_"))
  
  
  write_tsv(descending_data, paste("Maize_chr", i, "descending.txt", sep = "_"))
}
```

# Generate Teosinte 20 files
```{r}
for (i in 1:10) {
  print(paste("Processing Chromosome", i))
  ascending_data <- teosinte_snp_genotype %>% 
    filter(Chromosome == i) %>% 
    arrange(as.numeric(Position))
  descending_data <- teosinte_snp_genotype %>% 
    filter(Chromosome == i) %>% 
    arrange(desc(as.numeric(Position)))
  write_tsv(ascending_data, paste("teosinte_chr", i, "ascending.txt", sep = "_"))
  write_tsv(descending_data, paste("teosinte_chr", i, "descending.txt", sep = "_"))
}
```

# Data visualization
```{r}
# Load necessary libraries
library(ggplot2)
library(dplyr)
library(tidyr)
library(readr)

# Convert Chromosome and Position to numeric and remove missing values
maize_snp_genotype_clean <- maize_snp_genotype %>%
  mutate(Chromosome = as.numeric(Chromosome),
         Position = as.numeric(Position)) %>%
  drop_na()

teosinte_snp_genotype_clean <- teosinte_snp_genotype %>%
  mutate(Chromosome = as.numeric(Chromosome),
         Position = as.numeric(Position)) %>%
  drop_na()

# SNP Distribution Density per Chromosome for Maize
ggplot(data = maize_snp_genotype_clean) + 
  geom_density(mapping = aes(x = Position, fill = factor(Chromosome)), alpha = 0.6) + 
  facet_wrap(~factor(Chromosome), scales = "free_x") + 
  labs(title = "Maize SNP Distribution on Each Chromosome", 
       x = "Position on Chromosome", 
       y = "Density",
       fill = "Chromosome") +
  theme_minimal() +
  theme(plot.title = element_text(size = 16, face = "bold"),
        axis.title = element_text(size = 14))

# SNP Distribution Density per Chromosome for Teosinte
ggplot(data = teosinte_snp_genotype_clean) + 
  geom_density(mapping = aes(x = Position, fill = factor(Chromosome)), alpha = 0.6) + 
  facet_wrap(~factor(Chromosome), scales = "free_x") + 
  labs(title = "Teosinte SNP Distribution on Each Chromosome", 
       x = "Position on Chromosome", 
       y = "Density",
       fill = "Chromosome") +
  theme_minimal() +
  theme(plot.title = element_text(size = 16, face = "bold"),
        axis.title = element_text(size = 14))

# SNP Count per Chromosome for Maize and Teosinte
maize_snp_counts <- maize_snp_genotype_clean %>%
  group_by(Chromosome) %>%
  summarise(SNP_count = n())

teosinte_snp_counts <- teosinte_snp_genotype_clean %>%
  group_by(Chromosome) %>%
  summarise(SNP_count = n())

# Combine maize and teosinte SNP counts
snp_count_comparison <- bind_rows(
  maize_snp_counts %>% mutate(Source = "Maize"),
  teosinte_snp_counts %>% mutate(Source = "Teosinte")
)

# Bar plot for SNP counts per chromosome
ggplot(data = snp_count_comparison, aes(x = factor(Chromosome), y = SNP_count, fill = Source)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(title = "SNP Counts by Chromosome for Maize and Teosinte",
       x = "Chromosome",
       y = "SNP Count",
       fill = "Source") +
  scale_fill_manual(values = c("Maize" = "#E69F00", "Teosinte" = "#56B4E9")) +
  theme_minimal() +
  theme(plot.title = element_text(size = 16, face = "bold"),
        axis.title = element_text(size = 14),
        legend.title = element_text(size = 14))
```

# Spread and variability of SNP positions of each chromosome using a boxplot

```{r}
library(ggplot2)
maize_snp_genotype$Chromosome <- factor(
  maize_snp_genotype$Chromosome,
  levels = as.character(1:10),
  ordered = TRUE
)

ggplot(maize_snp_genotype, aes(x = Chromosome, y = as.numeric(Position), fill = Chromosome)) +
  geom_boxplot() +
  theme_minimal() +
  labs(x = "Chromosome", y = "Position", title = "Boxplot by Chromosome")
```

# Missing data and amount of homozygous,heterozygous and missing

```{r}

library(dplyr)
library(tidyr)
library(ggplot2)
library(tidyverse)

cleaned_fang <- fang_et_al_genotypes %>%
  select(-JG_OTU) %>%
  pivot_longer(cols = -c(Sample_ID, Group), names_to = "SNP_ID", values_to = "Sequence")

cleaned_fang <- cleaned_fang %>%
  mutate(new_sequence = case_when(
    Sequence %in% c("A/A", "T/T", "C/C", "G/G") ~ "Homozygous",
    Sequence == "?/?" ~ "Missing",
    TRUE ~ "Heterozygous"
  ))

ggplot(cleaned_fang, aes(x = new_sequence, fill = new_sequence)) +
  geom_bar() +
  theme_bw() +
  labs(title = "Distribution of Genotypes", x = "Genotype", y = "Count")
```
